File: int-mul.rb
def int_mul(x, y)
  if x < 10 && y < 10
    return x * y
  end

  digit_x = 0
  digit_y = 0
  x_dup = x
  y_dup = y
  while x_dup > 0
    x_dup = x_dup / 10
    digit_x += 1
  end
  while y_dup > 0
    y_dup = y_dup / 10
    digit_y += 1
  end
  division = ([digit_x, digit_y].max) / 2
  a = x / (10 ** division)
  b = y / (10 ** division)
  c = x % (10 ** division)
  d = y % (10 ** division)
  multi_1 = int_mul(a, b)
  multi_2 = int_mul(c, d)
  multi_3 = int_mul(a + c, b + d) - multi_1 - multi_2
  p multi_1
  p multi_2
  p multi_3
  result = multi_1 * (10 ** (2 * division)) + multi_2 + multi_3 * (10 ** division)
  p result
end

int_mul(3141592653589793238462643383279502884197169399375105820974944592, 2718281828459045235360287471352662497757247093699959574966967627)

File: week-1/int-mul.js
var integerMultiplication = (intX, intY) => {
  if (intX < 10 && intY < 10) {
    return intX * intY;
  }

  let digitX = 0, digitY = 0;
  let x = intX, y = intY;
  while (x / 10 > 0) {
    x = Math.floor(x / 10);
    digitX++;
  }
  while (y / 10 > 0) {
    y = Math.floor(y / 10);
    digitY++;
  }

  let n = Math.max(digitX, digitY);
  let half = Math.floor(n / 2);
  let a = Math.floor(intX / Math.pow(10, half)),
      b = Math.floor(intY / Math.pow(10, half)),
      c = intX % Math.pow(10, half),
      d = intY % Math.pow(10, half);
  let multi1 = integerMultiplication(a, b),
      multi2 = integerMultiplication(c, d),
      multi3 = integerMultiplication((a + c), (b + d)) - multi1 - multi2;

  let result = multi1 * Math.pow(10, 2 * half) + multi2 + multi3 * Math.pow(10, half);
  console.log(result)
  return result;
}

integerMultiplication(3141592653589793238462643383279502884197169399375105820974944592, 2718281828459045235360287471352662497757247093699959574966967627)

module.exports = integerMultiplication;